diff -urp udpxy.1/Makefile udpxy.full/Makefile
--- udpxy.1/Makefile	2012-03-16 11:05:51.000000000 +0400
+++ udpxy.full/Makefile	2012-03-11 13:34:00.000000000 +0400
@@ -99,7 +99,7 @@ INSTALLROOT := /usr/local
 ARCDIR = ..
 ARCFILE := $(ARCDIR)/$(EXEC).$(VERSION).$(BUILDNO)-$(PATCHNO)-$(BTYPE).tgz
 
-COMMON_OPT :=
+COMMON_OPT :=-D_GNU_SOURCE
 ifneq (yes,$(NO_UDPXREC))
 COMMON_OPT := $(COMMON_OPT) -DUDPXREC_MOD
 endif
diff -urp udpxy.1/osdef.h udpxy.full/osdef.h
--- udpxy.1/osdef.h	2012-03-16 11:07:05.000000000 +0400
+++ udpxy.full/osdef.h	2012-03-11 13:35:28.000000000 +0400
@@ -42,6 +42,9 @@
 #if defined(__linux)
     #define NO_SOCKADDR_SA_LEN
     #define USE_SELECT_READY
+    #ifdef _GNU_SOURCE
+        #define USE_ACCEPT4
+    #endif
 #endif
 
 #ifndef HAS_SOCKLEN
@@ -57,6 +60,9 @@
 #endif
 
 #if defined(__CYGWIN__)
+    #ifdef _GNU_SOURCE
+        #define USE_ACCEPT4
+    #endif
     #define NO_INET6_SUPPORT
     #define NO_SOCKADDR_SA_LEN
     #define NO_SO_RCVLOWAT
diff -urp udpxy.1/udpxy.c udpxy.full/udpxy.c
--- udpxy.1/udpxy.c	2012-03-16 11:06:10.000000000 +0400
+++ udpxy.full/udpxy.c	2012-03-11 13:34:00.000000000 +0400
@@ -983,7 +983,11 @@ accept_requests (int sockfd, tmfd_t* aso
     while (naccepted < nmax) {
         TRACE( (void)tmfputs ("Accepting new connection\n", g_flog) );
 
+#if defined(USE_ACCEPT4)
+        new_sockfd = accept4 (sockfd, (struct sockaddr*)&cliaddr, &addrlen, SOCK_NONBLOCK );
+#else
         new_sockfd = accept (sockfd, (struct sockaddr*)&cliaddr, &addrlen );
+#endif
         if (-1 == new_sockfd) {
             err = errno;
             if ((EWOULDBLOCK == err) || (EAGAIN == err)) {
@@ -1000,10 +1004,12 @@ accept_requests (int sockfd, tmfd_t* aso
             break;
         }
 
+#if !defined(USE_ACCEPT4)
         if (0 != set_nblock (new_sockfd, 1)) {
             (void) close (new_sockfd); /* TODO: error-aware close */
             continue;
         }
+#endif
         /*
         if (0 != set_timeouts(new_sockfd, new_sockfd, g_uopt.sr_tmout, 0,
             g_uopt.sw_tmout, 0)) {
